src/xarray_regrid/methods/flox_reduce.py

```diff
@@ -0,0 +1,174 @@
+"""Implementation of flox reduction based regridding methods."""
+
+from typing import overload
+
+import flox.xarray
+import numpy as np
+import pandas as pd
+import xarray as xr
+
+from xarray_regrid import utils
+from xarray_regrid.methods._shared import (
+    construct_intervals,
+    reduce_data_to_new_domain,
+    restore_properties,
+)
+
+
+@overload
+def statistic_reduce(
+    data: xr.DataArray,
+    target_ds: xr.Dataset,
+    time_dim: str | None,
+    method: str,
+    skipna: bool = False,
+) -> xr.DataArray: ...
+
+
+@overload
+def statistic_reduce(
+    data: xr.Dataset,
+    target_ds: xr.Dataset,
+    time_dim: str | None,
+    method: str,
+    skipna: bool = False,
+) -> xr.Dataset: ...
+
+
+def statistic_reduce(
+    data: xr.DataArray | xr.Dataset,
+    target_ds: xr.Dataset,
+    time_dim: str | None,
+    method: str,
+    skipna: bool = False,
+) -> xr.DataArray | xr.Dataset:
+    """Upsampling of data using statistical methods (e.g. the mean or variance).
+
+    We use flox Aggregations to perform a "groupby" over multiple dimensions, which we
+    reduce using the specified method.
+    https://flox.readthedocs.io/en/latest/aggregations.html
+
+    Args:
+        data: Input dataset.
+        target_ds: Dataset which coordinates the input dataset should be regrid to.
+        time_dim: Name of the time dimension. Defaults to "time". Use `None` to force
+            regridding over the time dimension.
+        method: One of the following reduction methods: "sum", "mean", "var", "std",
+            or "median.
+        skipna: If NaN values should be ignored.
+
+    Returns:
+        xarray.dataset with regridded land cover categorical data.
+    """
+    valid_methods = ["sum", "mean", "var", "std", "median"]
+    if method not in valid_methods:
+        msg = f"Invalid method. Please choose from '{valid_methods}'."
+        raise ValueError(msg)
+
+    if skipna:
+        method = "nan" + method
+
+    coords = utils.common_coords(data, target_ds, remove_coord=time_dim)
+    target_ds_sorted = target_ds.sortby(list(coords))
```

Comment:
This assumes `coords` are all "dimension coordinates". We could check `ds.indexes[coord].is_sorted` and just slice without sorting when `True`

Reply:
Good point!